experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp

```diff
@@ -196,6 +229,530 @@ struct PrintOpLowering : public ConversionPattern {
+    Value initialState = block->getArgument(0);
+    Value initialUpstreamState = rewriter.create<ExtractValueOp>(
+        op->getLoc(), upstreamInfo.stateType, initialState,
+        rewriter.getIndexArrayAttr({0}));
+
+    // Call Open on upstream.
+    func::CallOp openCallOp = rewriter.create<func::CallOp>(
+        op->getLoc(), upstreamInfo.openFunc, upstreamInfo.stateType,
+        initialUpstreamState);
+
+    // Update upstream state.
+    Value updatedState = rewriter.create<InsertValueOp>(
+        op->getLoc(), initialState, openCallOp->getResult(0),
+        rewriter.getIndexArrayAttr({0}));
+
+    // Return.
+    rewriter.create<func::ReturnOp>(op->getLoc(), updatedState);
+  }
+
+  // Create Next function. --------------------------------------------------
+  {
+    FuncOp funcOp = createNextFunc(op, rewriter, opInfo, elementType);
+    Block *block = &funcOp.getBody().front();
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(block);
+
+    // Extract upstream state.
+    Value initialState = block->getArgument(0);
+    Value initialUpstreamState = rewriter.create<ExtractValueOp>(
+        op->getLoc(), upstreamInfo.stateType, initialState,
+        rewriter.getIndexArrayAttr({0}));
+
+    // Get first result from upstream.
+    SmallVector<Type> nextResultTypes = {upstreamInfo.stateType,
+                                         rewriter.getI1Type(), elementType};
+    func::CallOp firstNextCall =
+        rewriter.create<func::CallOp>(op->getLoc(), upstreamInfo.nextFunc,
+                                      nextResultTypes, initialUpstreamState);
+
+    // Check for empty upstream.
+    scf::IfOp ifOp = rewriter.create<scf::IfOp>(op->getLoc(), nextResultTypes,
+                                                firstNextCall->getResult(1),
+                                                /*withElseRegion=*/true);
+
+    {
+      // Fill "then" region of ifOp.
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointToStart(&ifOp.getThenRegion().front());
+
+      // Create while loop.
+      SmallVector<Value> whileInputs = {firstNextCall->getResult(0),
+                                        firstNextCall->getResult(2)};
+      SmallVector<Type> whileInputTypes = {
+          firstNextCall->getResult(0).getType(),
+          firstNextCall->getResult(2).getType()};
+      SmallVector<Type> whileResultTypes = {upstreamInfo.stateType, elementType,
+                                            elementType};
+      SmallVector<Location> whileResultLocs = {op->getLoc(), op->getLoc(),
+                                               op->getLoc()};
+      scf::WhileOp whileOp = rewriter.create<scf::WhileOp>(
+          op->getLoc(), whileResultTypes, whileInputs);
+
+      // Fill the "before" region of whileOp.
+      {
+        OpBuilder::InsertionGuard guard(rewriter);
+        Block *before =
+            rewriter.createBlock(&whileOp.getBefore(), {}, whileInputTypes,
```

Comment:
We should have (or propose upstream) similar idiomatic builders for this.

Reply:
OK, I implemented a helper for myself for the time being in one of the latest commits.